File: rust/one_to_one.rs
use std::time::SystemTime;
use std::{thread, time};

mod liner;
use liner::Liner;

const MESS_SEND_COUNT: usize = 1000;
const MESS_SIZE: usize = 100;
const SEND_CYCLE_COUNT: usize = 10;

static mut RECEIVE_COUNT: i32 = 0;
static mut SEND_BEGIN: u64 = 0;
static mut SEND_END: u64 = 0;

fn cb1(_to: &str, _from: &str,  _data: &[u8]){
    println!("receive_from {}", _from);
}

fn cb2(_to: &str, _from: &str,  _data: &[u8]){
    unsafe {
        RECEIVE_COUNT += 1;

        if RECEIVE_COUNT == MESS_SEND_COUNT as i32{
            RECEIVE_COUNT = 0;
            println!("receive_from {} ms", current_time_ms() - SEND_END);
        }
    }
}

fn  main() {

    let mut client1 = Liner::new("client1", "topic_client1", "localhost:2255", "redis://localhost/", cb1);
    let mut client2 = Liner::new("client2", "topic_client2", "localhost:2256", "redis://localhost/", cb2);
   
    client1.clear_stored_messages();
    client2.clear_stored_messages();

    client1.clear_addresses_of_topic();
    client2.clear_addresses_of_topic();

    client1.run();
    client2.run();

    unsafe{
        let array = [0; MESS_SIZE];
        for _ in 0..SEND_CYCLE_COUNT{
            SEND_BEGIN = current_time_ms();
            for _ in 0..MESS_SEND_COUNT{
                client1.send_to("topic_client2", array.as_slice());
            }
            SEND_END = current_time_ms();
            println!("send_to {} ms", SEND_END - SEND_BEGIN);       
        
            thread::sleep(time::Duration::from_millis(1000));
        }
    }
}

fn current_time_ms()->u64{ 
    SystemTime::now()
    .duration_since(SystemTime::UNIX_EPOCH)
    .unwrap()
    .as_millis() as u64
}

File: cpp/one_to_one.cpp

#include "liner_broker.h"

#include <ctime>
#include <iostream>
#include <chrono>
#include <thread>

const int MESS_SEND_COUNT = 10000;
const int MESS_SIZE = 1024;
const int SEND_CYCLE_COUNT = 30;

int main(int argc, char* argv[])
{  
    auto client1 = LinerBroker("client1", "topic_client1", "localhost:2255", "redis://localhost/");
    auto client2 = LinerBroker("client2", "topic_client2", "localhost:2256", "redis://localhost/");
 
    int receive_count = 0;
    clock_t send_begin = clock();
    clock_t send_end = clock();

    client1.run([](const std::string& to, const std::string& from, const std::string& data){});
    client2.run([&receive_count, &send_end](const std::string& to, const std::string& from, const std::string& data){
        receive_count += 1;
        if (receive_count == MESS_SEND_COUNT){
            receive_count = 0;
            std::cout << "receive_from " << 1000.0 * (clock() - send_end) / CLOCKS_PER_SEC << " ms" << std::endl;
        }
    });
    
    char data[MESS_SIZE];
    for (int i = 0; i < SEND_CYCLE_COUNT; ++i){
        send_begin = clock();
        for (int j = 0; j < MESS_SEND_COUNT; ++j){
            client1.sendTo("topic_client2", data);
        }
        send_end = clock();
        std::cout << "send_to " << 1000.0 * (send_end - send_begin) / CLOCKS_PER_SEC << " ms" << std::endl;
        std::this_thread::sleep_for(std::chrono::milliseconds(1000));
    }    
}